### PR TITLE
Warn on implicit copies for event assignment within readonly members

### LINKED ISSUE #35610: WRN_ImplicitCopyInReadOnlyMember not given for event assignment
```cs
using System;
public struct S {
    public event Action E;
    public readonly void M()
    {
        E += M2; // should warn about implicit copy
    }
    
    public static void M2() { }
}
```

### PR DESCRIPTION
**Pull request by RikkiGibson**
Resolves #35610

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 1]
```diff
@@ -661,6 +661,73 @@ static void Main()
...
+        public void ReadOnlyMethod_CallEventAccessor()
+        {
+            var csharp = @"
+using System;
+
+public struct S
+{
+    public readonly void M()
+    {
+            Console.WriteLine(E is null);
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: >Console.WriteLine(E is null); [](start = 12, length = 29)

nit: unusual indent

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 2]
```diff
@@ -661,6 +661,73 @@ static void Main()
...
+            // should create local copy
+            E += () => {}; // warning
+            Console.WriteLine(E is null);
+            E -= () => {}; // warning
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: >E -= () => {}; // warning [](start = 12, length = 25)

Consider also testing direct assignment: `E = () => { };`

**Comment by RikkiGibson** [comment]
Need a second sign off for a small change please @dotnet/roslyn-compiler

**Comment by AlekseyTs** [comment]
@RikkiGibson It looks like some tests are failing

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 3]
```diff
@@ -262,6 +262,8 @@ private BoundExpression BindEventAssignment(AssignmentExpressionSyntax node, Bou
             }
             else
             {
+                CheckImplicitThisCopyInReadOnlyMember(receiverOpt, method, diagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: >CheckImplicitThisCopyInReadOnlyMember(receiverOpt, method, diagnostics); [](start = 16, length = 72)

It doesn't look like this is going to warn on an event of another struct stored in a field of the current type. Please make sure there is a test for this scenario

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 3)

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs (hunk 3 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: This is by design (unfortunately). See #33968. I can add a test to demonstrate the existing behavior though.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: Fixed

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: There's existing coverage for this. It's always an error within a readonly method. Will move event tests closer together so it's easier to find.

**Comment by RikkiGibson** [comment]
Oops, changing the whitespace broke the expected diagnostic locations. Will fix.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: > This is by design (unfortunately). See #33968. I can add a test to demonstrate the existing behavior though.

It looks like the issue is talking about a different scenario involving readonly fields in a non-readonly member

---
In reply to: [284909657](https://github.com/dotnet/roslyn/pull/35710#discussion_r284909657) [](ancestors = 284909657)

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: Let's sync up about this offline tomorrow.

**Comment by RikkiGibson** [comment]
@dotnet/roslyn-compiler Please review this small bug fix.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs [hunk 4]
```diff
@@ -760,7 +760,105 @@ public struct S
...
+{
+    public S2 s2;
+
+    public readonly void M()
+    {
+        s2.E += () => {};
+        s2.E -= () => {};
+    }
+
+    public readonly event Action E { add {} remove {} }
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs]: Nit: unecessary definiton, confused me for a moment.

**Pull request merged**